yaml to java conversion fails for ArrayList

### Issue description

Clone branch `security-tranining` at `micronaut-advocacy/micronaut-live-newsletter`

https://github.com/micronaut-advocacy/micronaut-live-newsletter/tree/security-training

run 

`./gradlew :app:prepareJitOptimizations`

```13:40:12.672 [main] INFO  i.m.a.s.s.YamlPropertySourceGenerator - Converting application.yml into Java based configuration
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at io.micronaut.aot.cli.Main.executeInIsolatedLoader(Main.java:118)
	at io.micronaut.aot.cli.Main.run(Main.java:92)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at io.micronaut.aot.cli.Main.execute(Main.java:139)
	at io.micronaut.aot.cli.Main.main(Main.java:143)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.micronaut.aot.cli.Main.executeInIsolatedLoader(Main.java:112)
	... 10 more
Caused by: java.lang.UnsupportedOperationException: Configuration map contains an entry of type class java.util.ArrayList which is not supported yet. Please file a bug report.
Caused by: java.lang.reflect.InvocationTargetException

	at io.micronaut.aot.std.sourcegen.MapPropertySourceGenerator.generateMap(MapPropertySourceGenerator.java:95)
	at io.micronaut.aot.std.sourcegen.MapPropertySourceGenerator.generate(MapPropertySourceGenerator.java:116)
	at io.micronaut.aot.core.codegen.AbstractSingleClassFileGenerator.generate(AbstractSingleClassFileGenerator.java:35)
	at io.micronaut.aot.std.sourcegen.YamlPropertySourceGenerator.createMapProperty(YamlPropertySourceGenerator.java:75)
	at io.micronaut.aot.std.sourcegen.YamlPropertySourceGenerator.generate(YamlPropertySourceGenerator.java:58)
Caused by: java.lang.UnsupportedOperationException: Configuration map contains an entry of type class java.util.ArrayList which is not supported yet. Please file a bug report.

	at io.micronaut.aot.std.sourcegen.AbstractStaticServiceLoaderSourceGenerator.generate(AbstractStaticServiceLoaderSourceGenerator.java:104)
	at io.micronaut.aot.core.codegen.ApplicationContextConfigurerGenerator.generate(ApplicationContextConfigurerGenerator.java:59)
	at io.micronaut.aot.MicronautAotOptimizer$Runner.execute(MicronautAotOptimizer.java:318)
	at io.micronaut.aot.MicronautAotOptimizer.execute(MicronautAotOptimizer.java:191)
	... 15 more

```

--- a/aot-std-optimizers/src/main/java/io/micronaut/aot/std/sourcegen/MapPropertySourceGenerator.java
+++ b/aot-std-optimizers/src/main/java/io/micronaut/aot/std/sourcegen/MapPropertySourceGenerator.java
@@ -105,9 +105,9 @@ public class MapPropertySourceGenerator extends AbstractSingleClassFileGenerator
                     prefix = "(short) ";
                 }
                 return prefix + format + appendix;
-            } else if (List.class.isAssignableFrom(valueClass) ) {
+            } else if (List.class.isAssignableFrom(valueClass)) {
                 return generateListMethod((List<?>) value, builder);
-            } else if (Map.class.isAssignableFrom(valueClass) ) {
+            } else if (Map.class.isAssignableFrom(valueClass)) {
                 return generateMapMethod((Map<?, ?>) value, builder);
             } else {
                 throw new UnsupportedOperationException("Configuration map contains an entry of type " + valueClass + " which is not supported yet. Please file a bug report.");

--- a/aot-std-optimizers/src/main/java/io/micronaut/aot/std/sourcegen/MapPropertySourceGenerator.java
+++ b/aot-std-optimizers/src/main/java/io/micronaut/aot/std/sourcegen/MapPropertySourceGenerator.java
@@ -28,11 +28,16 @@ import io.micronaut.core.annotation.NonNull;
 import io.micronaut.core.order.Ordered;
 import io.micronaut.core.util.StringUtils;
 
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
 import java.util.Map;
 
 import static io.micronaut.aot.std.sourcegen.MapPropertySourceGenerator.BASE_ORDER_OPTION;
+import static javax.lang.model.element.Modifier.PRIVATE;
 import static javax.lang.model.element.Modifier.PUBLIC;
+import static javax.lang.model.element.Modifier.STATIC;
 
 /**
  * A source generator which generates a map property source with a fixed
@@ -54,6 +59,7 @@ public class MapPropertySourceGenerator extends AbstractSingleClassFileGenerator
 
     private final String resourceName;
     private final Map<String, Object> values;
+    private int methodCount = 0;
 
     public MapPropertySourceGenerator(
             String resourceName,
@@ -62,43 +68,78 @@ public class MapPropertySourceGenerator extends AbstractSingleClassFileGenerator
         this.values = values;
     }
 
-    private CodeBlock generateMap() {
+    private CodeBlock generateMap(TypeSpec.Builder builder) {
         CodeBlock.Builder mapBuilder = CodeBlock.builder();
         mapBuilder.add("new $T() {{\n", HashMap.class);
         for (Map.Entry<String, Object> entry : values.entrySet()) {
             String key = entry.getKey();
             Object value = entry.getValue();
-            mapBuilder.add("put(\"" + key + "\", ");
-            if (value == null) {
-                mapBuilder.add("null");
-            } else {
-                Class<?> valueClass = value.getClass();
-                if (CharSequence.class.isAssignableFrom(valueClass)) {
-                    mapBuilder.add("\"" + value + "\"");
-                } else if (Number.class.isAssignableFrom(valueClass) || Boolean.class.isAssignableFrom(valueClass)) {
-                    String format = String.valueOf(value);
-                    String prefix = "";
-                    String appendix = "";
-                    if (Long.class.equals(valueClass)) {
-                        appendix = "L";
-                    } else if (Double.class.equals(valueClass)) {
-                        appendix = "D";
-                    } else if (Float.class.equals(valueClass)) {
-                        appendix = "F";
-                    } else if (Byte.class.equals(valueClass)) {
-                        prefix = "(byte) ";
-                    } else if (Short.class.equals(valueClass)) {
-                        prefix = "(short) ";
-                    }
-                    mapBuilder.add(prefix + format + appendix);
-                } else {
-                    throw new UnsupportedOperationException("Configuration map contains an entry of type " + valueClass + " which is not supported yet. Please file a bug report.");
-                }
-            }
+            mapBuilder.add("put(\"" + key + "\", " + convertValueToSource(value, builder));
             mapBuilder.add(");\n");
         }
         mapBuilder.add("}}");
         return mapBuilder.build();
+    }
+
+    private String convertValueToSource(Object value, TypeSpec.Builder builder) {
+        if (value == null) {
+             return "null";
+        } else {
+            Class<?> valueClass = value.getClass();
+            if (CharSequence.class.isAssignableFrom(valueClass)) {
+                return "\"" + value + "\"";
+            } else if (Number.class.isAssignableFrom(valueClass) || Boolean.class.isAssignableFrom(valueClass)) {
+                String format = String.valueOf(value);
+                String prefix = "";
+                String appendix = "";
+                if (Long.class.equals(valueClass)) {
+                    appendix = "L";
+                } else if (Double.class.equals(valueClass)) {
+                    appendix = "D";
+                } else if (Float.class.equals(valueClass)) {
+                    appendix = "F";
+                } else if (Byte.class.equals(valueClass)) {
+                    prefix = "(byte) ";
+                } else if (Short.class.equals(valueClass)) {
+                    prefix = "(short) ";
+                }
+                return prefix + format + appendix;
+            } else if (List.class.isAssignableFrom(valueClass) ) {
+                return generateListMethod((List<?>) value, builder);
+            } else if (Map.class.isAssignableFrom(valueClass) ) {
+                return generateMapMethod((Map<?, ?>) value, builder);
+            } else {
+                throw new UnsupportedOperationException("Configuration map contains an entry of type " + valueClass + " which is not supported yet. Please file a bug report.");
+            }
+        }
+    }
+
+    private String generateListMethod(List<?> value, TypeSpec.Builder builder) {
+        String methodName = "list" + methodCount++;
+        MethodSpec.Builder listMethod = MethodSpec.methodBuilder(methodName)
+                .addModifiers(PRIVATE, STATIC)
+                .returns(List.class);
+        listMethod.addStatement("$T result = new $T<>()", List.class, ArrayList.class);
+        for (Object o : value) {
+            listMethod.addStatement("result.add(" + convertValueToSource(o, builder) + ")");
+        }
+        listMethod.addStatement("return result");
+        builder.addMethod(listMethod.build());
+        return methodName + "()";
+    }
+
+    private String generateMapMethod(Map<?, ?> value, TypeSpec.Builder builder) {
+        String methodName = "map" + methodCount++;
+        MethodSpec.Builder mapMethod = MethodSpec.methodBuilder(methodName)
+                .addModifiers(PRIVATE, STATIC)
+                .returns(Map.class);
+        mapMethod.addStatement("$T result = new $T<>()", Map.class, LinkedHashMap.class);
+        for (Map.Entry<?, ?> entry : value.entrySet()) {
+            mapMethod.addStatement("result.put(" + convertValueToSource(entry.getKey(), builder) + ", " + convertValueToSource(entry.getValue(), builder) + ")");
+        }
+        mapMethod.addStatement("return result");
+        builder.addMethod(mapMethod.build());
+        return methodName + "()";
     }
 
     @Override
@@ -109,20 +150,19 @@ public class MapPropertySourceGenerator extends AbstractSingleClassFileGenerator
         int order = getContext().getConfiguration()
                 .optionalValue(orderKey, value ->
                         value.map(Integer::parseInt).orElse(Ordered.HIGHEST_PRECEDENCE));
-        TypeSpec type = TypeSpec.classBuilder(typeName)
+        TypeSpec.Builder typeBuilder = TypeSpec.classBuilder(typeName)
                 .addModifiers(PUBLIC)
-                .superclass(MapPropertySource.class)
-                .addMethod(MethodSpec.constructorBuilder()
-                        .addStatement("super($S, $L)", resourceName, generateMap())
+                .superclass(MapPropertySource.class);
+        typeBuilder.addMethod(MethodSpec.constructorBuilder()
+                        .addStatement("super($S, $L)", resourceName, generateMap(typeBuilder))
                         .build())
                 .addMethod(MethodSpec.methodBuilder("getOrder")
                         .addModifiers(PUBLIC)
                         .returns(int.class)
                         .addStatement("return $L", order)
                         .build())
-                .addAnnotation(Generated.class)
-                .build();
-        return javaFile(type);
+                .addAnnotation(Generated.class);
+        return javaFile(typeBuilder.build());
     }
 
     private String computeTypeName() {

--- a/aot-std-optimizers/src/main/java/io/micronaut/aot/std/sourcegen/MapPropertySourceGenerator.java
+++ b/aot-std-optimizers/src/main/java/io/micronaut/aot/std/sourcegen/MapPropertySourceGenerator.java
@@ -29,6 +29,7 @@ import io.micronaut.core.order.Ordered;
 import io.micronaut.core.util.StringUtils;
 
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.List;
@@ -119,11 +120,17 @@ public class MapPropertySourceGenerator extends AbstractSingleClassFileGenerator
         MethodSpec.Builder listMethod = MethodSpec.methodBuilder(methodName)
                 .addModifiers(PRIVATE, STATIC)
                 .returns(List.class);
-        listMethod.addStatement("$T result = new $T<>()", List.class, ArrayList.class);
-        for (Object o : value) {
-            listMethod.addStatement("result.add(" + convertValueToSource(o, builder) + ")");
+        if (value.isEmpty()) {
+            listMethod.addStatement("return $.emptyList()", Collections.class);
+        } else if (value.size() == 1) {
+            listMethod.addStatement("return $T.singletonList($L)", Collections.class, convertValueToSource(value.get(0), builder));
+        } else {
+            listMethod.addStatement("$T result = new $T<>($L)", List.class, ArrayList.class, value.size());
+            for (Object o : value) {
+                listMethod.addStatement("result.add(" + convertValueToSource(o, builder) + ")");
+            }
+            listMethod.addStatement("return result");
         }
-        listMethod.addStatement("return result");
         builder.addMethod(listMethod.build());
         return methodName + "()";
     }
@@ -133,11 +140,18 @@ public class MapPropertySourceGenerator extends AbstractSingleClassFileGenerator
         MethodSpec.Builder mapMethod = MethodSpec.methodBuilder(methodName)
                 .addModifiers(PRIVATE, STATIC)
                 .returns(Map.class);
-        mapMethod.addStatement("$T result = new $T<>()", Map.class, LinkedHashMap.class);
-        for (Map.Entry<?, ?> entry : value.entrySet()) {
-            mapMethod.addStatement("result.put(" + convertValueToSource(entry.getKey(), builder) + ", " + convertValueToSource(entry.getValue(), builder) + ")");
+        if (value.isEmpty()) {
+            mapMethod.addStatement("return $.emptyMap()", Collections.class);
+        } else if (value.size() == 1) {
+            Map.Entry<?, ?> entry = value.entrySet().iterator().next();
+            mapMethod.addStatement("return $T.singletonMap($L, $L)", Collections.class, convertValueToSource(entry.getKey(), builder), convertValueToSource(entry.getValue(), builder));
+        } else {
+            mapMethod.addStatement("$T result = new $T<>($L)", Map.class, LinkedHashMap.class, value.size());
+            for (Map.Entry<?, ?> entry : value.entrySet()) {
+                mapMethod.addStatement("result.put(" + convertValueToSource(entry.getKey(), builder) + ", " + convertValueToSource(entry.getValue(), builder) + ")");
+            }
+            mapMethod.addStatement("return result");
         }
-        mapMethod.addStatement("return result");
         builder.addMethod(mapMethod.build());
         return methodName + "()";
     }

--- a/aot-std-optimizers/src/test/groovy/io/micronaut/aot/std/sourcegen/YamlPropertySourceGeneratorTest.groovy
+++ b/aot-std-optimizers/src/test/groovy/io/micronaut/aot/std/sourcegen/YamlPropertySourceGeneratorTest.groovy
@@ -39,6 +39,7 @@ class YamlPropertySourceGeneratorTest extends AbstractSourceGeneratorSpec {
 import io.micronaut.context.env.MapPropertySource;
 import io.micronaut.core.annotation.Generated;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.List;
@@ -56,13 +57,11 @@ public class Test_configStaticPropertySource extends MapPropertySource {
   }
 
   private static List list2() {
-    List result = new ArrayList<>();
-    result.add("isAnonymous()");
-    return result;
+    return Collections.singletonList("isAnonymous()");
   }
 
   private static Map map1() {
-    Map result = new LinkedHashMap<>();
+    Map result = new LinkedHashMap<>(3);
     result.put("pattern", "/api/v1/subscriber/confirm");
     result.put("http-method", "PATCH");
     result.put("access", list2());
@@ -70,13 +69,11 @@ public class Test_configStaticPropertySource extends MapPropertySource {
   }
 
   private static List list4() {
-    List result = new ArrayList<>();
-    result.add("isAnonymous()");
-    return result;
+    return Collections.singletonList("isAnonymous()");
   }
 
   private static Map map3() {
-    Map result = new LinkedHashMap<>();
+    Map result = new LinkedHashMap<>(3);
     result.put("pattern", "/api/v1/subscriber/count");
     result.put("http-method", "GET");
     result.put("access", list4());
@@ -84,7 +81,7 @@ public class Test_configStaticPropertySource extends MapPropertySource {
   }
 
   private static List list0() {
-    List result = new ArrayList<>();
+    List result = new ArrayList<>(2);
     result.add(map1());
     result.add(map3());
     return result;

--- a/aot-std-optimizers/src/test/groovy/io/micronaut/aot/std/sourcegen/YamlPropertySourceGeneratorTest.groovy
+++ b/aot-std-optimizers/src/test/groovy/io/micronaut/aot/std/sourcegen/YamlPropertySourceGeneratorTest.groovy
@@ -38,7 +38,11 @@ class YamlPropertySourceGeneratorTest extends AbstractSourceGeneratorSpec {
 
 import io.micronaut.context.env.MapPropertySource;
 import io.micronaut.core.annotation.Generated;
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
 
 @Generated
 public class Test_configStaticPropertySource extends MapPropertySource {
@@ -47,7 +51,43 @@ public class Test_configStaticPropertySource extends MapPropertySource {
         put("micronaut.application.name", "demoApp");
         put("micronaut.server.port", 8181);
         put("micronaut.server.cors.enabled", true);
+        put("micronaut.security.intercept-url-map", list0());
         }});
+  }
+
+  private static List list2() {
+    List result = new ArrayList<>();
+    result.add("isAnonymous()");
+    return result;
+  }
+
+  private static Map map1() {
+    Map result = new LinkedHashMap<>();
+    result.put("pattern", "/api/v1/subscriber/confirm");
+    result.put("http-method", "PATCH");
+    result.put("access", list2());
+    return result;
+  }
+
+  private static List list4() {
+    List result = new ArrayList<>();
+    result.add("isAnonymous()");
+    return result;
+  }
+
+  private static Map map3() {
+    Map result = new LinkedHashMap<>();
+    result.put("pattern", "/api/v1/subscriber/count");
+    result.put("http-method", "GET");
+    result.put("access", list4());
+    return result;
+  }
+
+  private static List list0() {
+    List result = new ArrayList<>();
+    result.add(map1());
+    result.add(map3());
+    return result;
   }
 
   public int getOrder() {

--- a/aot-std-optimizers/src/test/resources/test-config.yml
+++ b/aot-std-optimizers/src/test/resources/test-config.yml
@@ -5,3 +5,15 @@ micronaut:
     port: 8181
     cors:
       enabled: true
+  security:
+    intercept-url-map:
+      -
+        pattern: /api/v1/subscriber/confirm
+        http-method: PATCH
+        access:
+          - isAnonymous()
+      -
+        pattern: /api/v1/subscriber/count
+        http-method: GET
+        access:
+          - isAnonymous()

